File: Algorithm and Data Structure/Baekjoon/12886번 돌 그룹/boj12886_rebas.cc
#include <cstdio>
#include <queue>
#include <utility>
#include <algorithm>
using namespace std;

int A, B, C, D;
bool check[1000][1000];

void bfs() {
    queue<pair<int, int>> q;
    q.push(make_pair(A, B));
    check[A][B] = true;
    while (!q.empty()) {
        int x = q.front().first, y = q.front().second; q.pop();
        int z = D-x-y;
        if (x == y && y == z) {
            printf("1\n");
            return;
        }
        int nx[] = {x, x, y}, ny[] = {y, z, z};
        for (int i=0; i<3; i++) {
            int a = nx[i], b = ny[i];
            if (a < b) b -= a, a += a;
            else if (a > b) a -= b, b += b;
            else continue;
            int c = D-a-b;
            int X = min(min(a, b), c);
            int Y = max(max(a, b), c);
            if (!check[X][Y]) {
                q.push(make_pair(X, Y));
                check[X][Y] = true;
            }
        }
    }
    printf("0\n");
}

void solve() {
    if (D%3) {
        printf("0\n");
        return;
    } else bfs();
}

int main() {
    scanf("%d %d %d", &A, &B, &C);
    D = A+B+C;
    solve();
    return 0;
}

File: Algorithm and Data Structure/Baekjoon/12886번 돌 그룹/boj12886_rebas.py
from collections import deque

def bfs():
    q = deque()
    q.append((A, B))
    check[A][B] = True
    while q:
        x, y = q.popleft()
        z = D-x-y
        if x == y == z:
            print(1)
            return
        for a, b in (x, y), (x, z), (y, z):
            if a < b:
                b -= a
                a += a
            elif a > b:
                a -= b
                b += b
            else:
                continue
            c = D-a-b
            X = min(a, b, c)
            Y = max(a, b, c)
            if not check[X][Y]:
                q.append((X, Y))
                check[X][Y] = True
    print(0)

def solve():
    if D % 3:
        print(0)
        return
    else:
        bfs()

A, B, C = map(int, input().split())
D = A+B+C
check = [[False]*D for _ in range(D)]
solve()


